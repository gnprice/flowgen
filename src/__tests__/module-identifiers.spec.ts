import { compiler, beautify } from "..";
import "../test-matchers";

Error.stackTraceLimit = 20;

it("should handle react types", () => {
  const ts = `
import type {ReactNode, ReactElement, RefAttributes} from 'react'
import * as React from 'react'
declare function s(node: ReactNode): void;
declare function s(node: React.ReactNode): void;
declare function s(node: ReactElement<'div'>): void;
declare function s(node: React.ReactElement<'div'>): void;

type A = React.RefAttributes<number>;
type AA = RefAttributes<number>;
type B = React.RefAttributes<React.Component<{ x: string }>>;
type BB = RefAttributes<React.Component<{ x: string }>>;
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

describe("should handle global types", () => {
  test("jsx", () => {
    const ts = `
import * as React from 'react'
declare function s(node: JSX.Element): void;

type Props = {children: JSX.Element}

declare class Component extends React.Component<Props> {
  render(): JSX.Element
}
`;
    const result = compiler.compileDefinitionString(ts, { quiet: true });
    expect(beautify(result)).toMatchSnapshot();
    expect(result).toBeValidFlowTypeDeclarations();
  });
});
