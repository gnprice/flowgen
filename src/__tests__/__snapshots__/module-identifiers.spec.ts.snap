// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle global types jsx 1`] = `
"import * as React from \\"react\\";
declare function s(node: React$Node): void;
declare type Props = {
  children: React$Node,
  ...
};
declare class Component mixins React.Component<Props> {
  render(): React$Node;
}
"
`;

exports[`should handle react types pure renames 1`] = `
"import type { Node, Element } from \\"react\\";
import * as React from \\"react\\";
declare function s(node: Node): void;
declare function s(node: React.Node): void;
declare function s(node: Element<\\"div\\">): void;
declare function s(node: React.Element<\\"div\\">): void;
"
`;

exports[`should handle react types rewrite away React.RefAttributes 1`] = `
"import type { Node } from \\"react\\";
import type {} from \\"react\\";
import * as React from \\"react\\";
import * as RenamedAct from \\"react\\";
declare class C {}
declare type A = {
  +ref?: number,
  ...
};
declare type AA = {
  +ref?: number,
  ...
};
declare type B = {
  +ref?: C,
  ...
};
declare type BB = {
  +ref?: C,
  ...
};
"
`;
